modules/dnn/src/layers/max_unpooling_layer.cpp

```diff
@@ -43,13 +43,18 @@ class MaxUnpoolLayerImpl CV_FINAL : public MaxUnpoolLayer
                          std::vector<MatShape> &outputs,
                          std::vector<MatShape> &internals) const CV_OVERRIDE
     {
-        CV_Assert(inputs.size() == 2);
+        CV_Assert(inputs.size() == 2 || inputs.size() == 3);
         CV_Assert(total(inputs[0]) == total(inputs[1]));
 
         MatShape outShape = inputs[0];
         outShape[2] = (outShape[2] - 1) * poolStride.height + poolKernel.height - 2 * poolPad.height;
         outShape[3] = (outShape[3] - 1) * poolStride.width + poolKernel.width - 2 * poolPad.width;
 
+        if (inputs.size() == 3)
+        {
+            outShape = inputs[2];
+        }
+
```

Comment:
That's good decision. I also couldn't find any other alternatives.

So let's rewrite it in this way:

```cpp
MatShape outShape;
if (inputs.size() == 2)
{
  outShape = inputs[0];
  outShape[2] = (outShape[2] - 1) * poolStride.height + poolKernel.height - 2 * poolPad.height;
  outShape[3] = (outShape[3] - 1) * poolStride.width + poolKernel.width - 2 * poolPad.width;
}
else
  outShape = inputs[2];

outputs.clear();
outputs.push_back(outShape);
```